### PR TITLE
[X86] Avoid useless DomTree in flags copy lowering

### PR DESCRIPTION
**Pull request by aengelke**
Currently, flags copy lowering does two expensive things:

- It traverses the CFG in RPO, and
- It requires a dominator tree that is not preserved.

Most notably, it is the only machine dominator tree user at -O0.

Many functions have no flag copies to begin with, therefore, add an early exit if EFLAGS has no COPY def.

The legacy pass manager has no way to dynamically decide whether an analysis is required. Therefore, if there's a copy, get the dominator tree from the pass manager, if it has one, otherwise, compute it.

These changes should make the pass very cheap for the common case.

### CODE REVIEW

File: llvm/lib/Target/X86/X86FlagsCopyLowering.cpp [hunk 1]
```diff
@@ -258,13 +258,38 @@ bool X86FlagsCopyLoweringPass::runOnMachineFunction(MachineFunction &MF) {
...
   TRI = Subtarget->getRegisterInfo();
-  MDT = &getAnalysis<MachineDominatorTreeWrapperPass>().getDomTree();
   PromoteRC = &X86::GR8RegClass;
 
   if (MF.empty())
     // Nothing to do for a degenerate empty function...
     return false;
 
+  bool HasCopies = false;
+  for (const MachineInstr &DefInst : MRI->def_instructions(X86::EFLAGS)) {
```
Comment (topperc) [inline llvm/lib/Target/X86/X86FlagsCopyLowering.cpp:268]: Can we use `llvm::none_of` with a lambda?

**Comment by topperc** [comment]
Shouldn't the O0-pipeline.ll test fail since the dominator tree is no longer needed at -O0?

**Comment by aengelke** [comment]
> Shouldn't the O0-pipeline.ll test fail since the dominator tree pass is no longer needed at -O0?

As a matter of fact they do, and I didn't notice, because these tests only runs in an assert build (do you know why it's not run in release builds?). Fixed now.

**Comment by KanRobert** [comment]
> > Shouldn't the O0-pipeline.ll test fail since the dominator tree pass is no longer needed at -O0?
> 
> As a matter of fact they do, and I didn't notice, because these tests only runs in an assert build (do you know why it's not run in release builds?). Fixed now.

`--debug-pass` does not work when `LLVM_ENABLE_ASSERTIONS` is off.

**Comment by aengelke** [comment]
> `--debug-pass` does not work when `LLVM_ENABLE_ASSERTIONS` is off.

It just happened to work for me and it also works with Fedora's build of LLVM (I'm certain they build without assertions).

File: llvm/lib/Target/X86/X86FlagsCopyLowering.cpp [hunk 2]
```diff
@@ -258,13 +258,32 @@ bool X86FlagsCopyLoweringPass::runOnMachineFunction(MachineFunction &MF) {
...
   TRI = Subtarget->getRegisterInfo();
-  MDT = &getAnalysis<MachineDominatorTreeWrapperPass>().getDomTree();
   PromoteRC = &X86::GR8RegClass;
 
   if (MF.empty())
     // Nothing to do for a degenerate empty function...
     return false;
 
+  if (none_of(MRI->def_instructions(X86::EFLAGS), [](const MachineInstr &MI) {
+        return MI.getOpcode() == TargetOpcode::COPY;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86FlagsCopyLowering.cpp:268]: Why don't check `MI.getOperand(0).getReg() == X86::EFLAGS` as well?

File: llvm/lib/Target/X86/X86FlagsCopyLowering.cpp [hunk 3]
```diff
@@ -688,7 +707,7 @@ bool X86FlagsCopyLoweringPass::runOnMachineFunction(MachineFunction &MF) {
       }
 #endif
 
-  return true;
+  return !Copies.empty();
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86FlagsCopyLowering.cpp:710]: Shouldn't we always return `true` with checking `MI.getOperand(0).getReg() == X86::EFLAGS`?

File: llvm/lib/Target/X86/X86FlagsCopyLowering.cpp (hunk 2 above)
Comment (aengelke) [reply llvm/lib/Target/X86/X86FlagsCopyLowering.cpp:268]: I don't think that's necessary (or beneficial): if there's a def of eflags from a copy instruction, then the first operand must be eflags.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**